polar-core/src/polar.rs

```diff
@@ -183,19 +182,19 @@ impl Polar {
                         let mut rule_warnings = check_singletons(&rule, &*kb)?;
                         warnings.append(&mut rule_warnings);
                         let rule = rewrite_rule(rule, kb);
-
-                        let name = rule.name.clone();
-                        let generic_rule = kb
-                            .rules
-                            .entry(name.clone())
-                            .or_insert_with(|| GenericRule::new(name, vec![]));
-                        generic_rule.add_rule(Arc::new(rule));
+                        kb.add_rule(rule);
                     }
                     parser::Line::Query(term) => {
                         kb.inline_queries.push(term);
                     }
+                    parser::Line::RuleTemplate(prototype) => {
+                        let prototype = rewrite_rule(prototype, kb);
```

Comment:
Why do we need to rewrite prototypes?

Reply:
I don't know, there's no point since we never check the body but I wasn't sure if there's a case where a head-only rule would get rewritten.

Reply:
Rewrites always go into the body, so we shouldn't need them. They're used for operations that return a value, like `dot`, `add`, `sub`, etc. Imo none of these should be allowed in rule prototypes, but I don't think we'd actually catch them.

I guess we could run the rewrites and then error if the rule has a body afterwards? Which would mean it's an invalid prototype also?

Reply:
Yeah we could do that